grid-proxy/internal/gpuindexer/gpuindexer.go

```diff
@@ -0,0 +1,176 @@
+
+type NodeGPUIndexer struct {
+	db              db.Database
+	relayClient     *direct.Client
+	nodesGPUChan    chan []types.NodeGPU
+	checkInterval   time.Duration
+	workerBatchSize int
+}
+
+func NewNodeGPUIndexer(
+	ctx context.Context,
+	relayURL,
+	mnemonics string,
+	sub *substrate.Substrate,
+	db db.Database,
+	indexerCheckIntervalMins int,
+	workerBatchSize int) (*NodeGPUIndexer, error) {
+	indexer := &NodeGPUIndexer{
+		db:              db,
+		nodesGPUChan:    make(chan []types.NodeGPU),
+		checkInterval:   time.Duration(indexerCheckIntervalMins) * time.Minute,
+		workerBatchSize: workerBatchSize,
+	}
+
+	client, err := direct.NewClient(ctx, direct.KeyTypeSr25519, mnemonics, relayURL, "tfgrid_proxy_indexer", sub, true, indexer.relayCallback)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create direct RMB client: %w", err)
+	}
+	indexer.relayClient = client
+
+	return indexer, nil
+}
+
+func (n *NodeGPUIndexer) worker(ctx context.Context) {
+	var nodesGPUBuffer []types.NodeGPU
+	ticker := time.NewTicker(workerCleanupInterval)
+	for {
+		select {
+		case nodesGPU := <-n.nodesGPUChan:
+			nodesGPUBuffer = append(nodesGPUBuffer, nodesGPU...)
+			if len(nodesGPUBuffer) >= n.workerBatchSize {
+				log.Debug().Msg("flushing gpu indexer buffer")
+				err := n.db.UpsertNodesGPU(nodesGPUBuffer)
+				if err != nil {
+					log.Error().Err(err).Msg("failed to update GPU info in GPU indexer")
+					continue
+				}
+				nodesGPUBuffer = nil
+			}
+		// This case should only be triggered when there leftovers data in the buffer
+		case <-ticker.C:
+			if len(nodesGPUBuffer) != 0 {
+				log.Debug().Msg("cleaning up gpu indexer buffer")
+				err := n.db.UpsertNodesGPU(nodesGPUBuffer)
+				if err != nil {
+					log.Error().Err(err).Msg("failed to update GPU info in GPU indexer")
+					continue
+				}
+				nodesGPUBuffer = nil
+			}
+		case <-ctx.Done():
+			log.Error().Err(ctx.Err()).Msg("worker exited")
+			return
+
+		}
+	}
+}
+
+func (n *NodeGPUIndexer) relayCallback(response *rmbTypes.Envelope, callBackErr error) {
+
+	errResp := response.GetError()
+
+	if errResp != nil {
+		log.Error().Msg(errResp.Message)
+		return
+	}
+
+	resp := response.GetResponse()
+	if resp == nil {
+		log.Error().Msg("received a non response envelope")
+		return
+	}
+
+	if response.Schema == nil || *response.Schema != rmb.DefaultSchema {
+		log.Error().Msgf("invalid schema received expected '%s'", rmb.DefaultSchema)
+		return
+	}
+
+	output := response.Payload.(*rmbTypes.Envelope_Plain).Plain
+
+	var nodesGPU []types.NodeGPU
+	err := json.Unmarshal(output, &nodesGPU)
+	if err != nil {
+		log.Error().Err(err).Msg("failed to unmarshal GPU information response")
+		return
+
+	}
+	for i := range nodesGPU {
+		nodesGPU[i].NodeTwinID = response.Source.Twin
+	}
+	// Will be using only one worker giving the current amount of nodes supporting GPUs
+	if len(nodesGPU) != 0 {
+		n.nodesGPUChan <- nodesGPU
+	}
+
+}
+
+func (n *NodeGPUIndexer) Start(ctx context.Context) {
```

Comment:
I'd do the following here

```
go queryGridNodes()
go gpuNodeResultsBatcher()
go gpuBatchesDBUpserter()
```



queryGridNodes (called `run`), this is being ran to index the gridNodes maybe every 1hr? could be less, I'm thinking also of spliting worker to reflect the behavior into two (gpuNodeResultsBatcher, gpuBatchesDBUpserter) where we separate the concerns of preparing batches and inserting them into the database 

I probably mislead you on Line:120 `// Will be using only one worker giving the current amount of nodes supporting GPUs` as we should never assume such, we should assume the other way around, the grid will have e.g 4000 nodes with GPUs

So, There's a goroutine that queries the nodes regularly, hands over the results of each node to the nodesGPUChan, then there's one or more workers creating batches, and there's one or more workers upserting the batches into the database